alerting/src/main/kotlin/org/opensearch/alerting/alerts/AlertMover.kt

```diff
@@ -22,97 +27,221 @@ import org.opensearch.common.xcontent.XContentHelper
+                throw RuntimeException(
+                    "Failed to delete alerts for [$monitorId, ${monitor?.triggers?.map { it.id }}]: " +
+                        deleteResponse.buildFailureMessage(),
+                    retryCause
+                )
+            }
+        }
+
+        private fun alertContentParser(bytesReference: BytesReference): XContentParser {
+            val xcp = XContentHelper.createParser(
+                NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE,
+                bytesReference, XContentType.JSON
             )
-            .version(hit.version)
-            .versionType(VersionType.EXTERNAL_GTE)
-            .id(hit.id)
-    }
-    val copyRequest = BulkRequest().add(indexRequests)
-    val copyResponse: BulkResponse = client.suspendUntil { bulk(copyRequest, it) }
-
-    val deleteRequests = copyResponse.items.filterNot { it.isFailed }.map {
-        DeleteRequest(alertIndex, it.id)
-            .routing(monitorId)
-            .version(it.version)
-            .versionType(VersionType.EXTERNAL_GTE)
-    }
-    val deleteResponse: BulkResponse = client.suspendUntil { bulk(BulkRequest().add(deleteRequests), it) }
-
-    if (copyResponse.hasFailures()) {
-        val retryCause = copyResponse.items.filter { it.isFailed }
-            .firstOrNull { it.status() == RestStatus.TOO_MANY_REQUESTS }
-            ?.failure?.cause
-        throw RuntimeException(
-            "Failed to copy alerts for [$monitorId, ${monitor?.triggers?.map { it.id }}]: " +
-                copyResponse.buildFailureMessage(),
-            retryCause
-        )
-    }
-    if (deleteResponse.hasFailures()) {
-        val retryCause = deleteResponse.items.filter { it.isFailed }
-            .firstOrNull { it.status() == RestStatus.TOO_MANY_REQUESTS }
-            ?.failure?.cause
-        throw RuntimeException(
-            "Failed to delete alerts for [$monitorId, ${monitor?.triggers?.map { it.id }}]: " +
-                deleteResponse.buildFailureMessage(),
-            retryCause
-        )
-    }
-}
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, xcp.nextToken(), xcp)
+            return xcp
+        }
+
+        /**
+         * Moves defunct active alerts to the alert history index when the corresponding workflow or trigger is deleted.
+         *
+         * The logic for moving alerts consists of:
+         * 1. Find active alerts:
+         *      a. matching workflowId if no workflow is provided (postDelete)
+         *      b. matching workflowid and chained alert triggerIds if monitor is provided (postIndex)
+         * 2. Move alerts over to DataSources.alertsHistoryIndex as DELETED
+         * 3. Delete alerts from monitor's DataSources.alertsIndex
+         * 4. Schedule a retry if there were any failures
+         */
+        suspend fun moveAlerts(client: Client, workflowId: String, workflow: Workflow?, monitorCtx: MonitorRunnerExecutionContext) {
```

Comment:
Can we ensure we are consistent and not adding regression during version upgrades. Two important things to consider:
1. Serialization/De-Ser during node to node interaction (with old and new version of the plugin) is not breaking and new version is backward compatible. 
2. Data (History, Config and System/Hidden indices) are read/updated across versions without causing regressions.

Reply:
Yes. There won't be regressions.
1. The moveAlerts method is always called within a try catch block and will not cause any issue with cross version Serialization/De-Ser.
2. We have added workflows as a new resource and have not edited any current resource.
3. all fields related to workflows have been

Reply:
Thanks, Can we ensure all transport interactions are also adhering to it and are backward compatible. This is to ensure no regressions due to heterogeneous fleet during upgrades.